src/core/Akka.Streams/Implementation/Sources.cs

```diff
@@ -130,20 +132,22 @@ public void OnPull()
                         }
                     }
                 }
-                else if (_buffer.NonEmpty)
+                else
                 {
-                    Push(_stage.Out, _buffer.Dequeue());
-                    if (_pendingOffer != null)
+                    if (_buffer.NonEmpty)
                     {
-                        EnqueueAndSuccess(_pendingOffer);
-                        _pendingOffer = null;
+                        Push(_stage.Out, _buffer.Dequeue());
+                        if (_pendingOffer != null)
+                        {
+                            EnqueueAndSuccess(_pendingOffer);
+                            _pendingOffer = null;
+                        }
+                    }
+                    if (_terminating && _buffer.IsEmpty)
+                    {
+                        _completion.SetResult(new object());
+                        CompleteStage();
                     }
-                }
-
-                if (_terminating && _buffer.IsEmpty)
```

Comment:
This is an actual bug, `_buffer` can be null if `_stage._maxBuffer` is less than 1. Moved the code to inside the previous if block